sql/core/src/main/scala/org/apache/spark/sql/DataFrameWriter.scala

```diff
@@ -383,14 +383,19 @@ final class DataFrameWriter[T] private[sql](ds: Dataset[T]) {
         // Get all input data source relations of the query.
         val srcRelations = df.logicalPlan.collect {
           case LogicalRelation(src: BaseRelation, _, _) => src
+          case relation: CatalogRelation if DDLUtils.isHiveTable(relation.catalogTable) => relation
         }
         EliminateSubqueryAliases(catalog.lookupRelation(tableIdentWithDB)) match {
           // Only do the check if the table is a data source table (the relation is a BaseRelation).
-          // TODO(cloud-fan): also check hive table relation here when we support overwrite mode
           // for creating hive tables.
           case LogicalRelation(dest: BaseRelation, _, _) if srcRelations.contains(dest) =>
```

Comment:
We only check `BaseRelation`. Let me check whether we can get rid of the comparison of `output` from the `equals` of `LogicalRelation` and see whether we can hit test case failure.